packages/cozy-stack-client/src/AppCollection.js

```diff
@@ -30,13 +35,47 @@ class AppCollection extends DocumentCollection {
       )
       id = idArg
     }
-    return Collection.get(
-      this.stackClient,
-      `${this.endpoint}${encodeURIComponent(id)}`,
-      {
-        normalize: this.constructor.normalizeDoctype(this.doctype)
+
+    if (!query || !query.sources || query.sources.length < 1) {
+      return Collection.get(
+        this.stackClient,
+        `${this.endpoint}${encodeURIComponent(id)}`,
+        {
+          normalize: this.constructor.normalizeDoctype(this.doctype)
+        }
+      )
+    }
+
+    if (!Array.isArray(query.sources)) {
+      throw new Error('query.sources must be an array')
+    }
+
+    const dataFetchers = {
+      stack: () => this.stackClient.fetchJSON('GET', this.endpoint + id),
+      registry: () => this.stackClient.fetchJSON('GET', registryEndpoint + id)
+    }
+
+    for (const source of query.sources) {
+      try {
+        const res = await dataFetchers[source]()
+
+        const data =
+          source === 'registry'
+            ? transformRegistryFormatToStackFormat(res)
+            : res.data
+
+        const normalizedDoc = normalizeDoc(data, this.doctype)
+
+        return { data: [normalizedDoc] }
+      } catch (err) {
+        if (
+          err.status === 404 ||
```

Comment:
Mmh I do not understand the logic. If a konnector is not installed, the response will be a 404 from the stack ? Is this not the case where we want to retrieve the konnector from the registry ? It seems like the logic should be reversed: if we have a 404, we "continue" and try to call the registry.

Reply:
La logique est de se dire que si c'est un 404 ça part en erreur aussi sur registry donc pas la peine de continuer. Ce n'est pas le cas ?

Reply:
Non, si ca 404 au premier appel c'est que l'utilisateur n'a pas le connecteur installé. Et donc derrière sur le registry ca va sûrement fonctionner.

Sur le même sujet de logique, je vois "Si la stack ne retourne rien, on souhaite alors interroger la registry pour récupérer ces données." dans la description de la PR. Est ce que tu peux éditer la description pour la rendre plus précise sur ce quoi on implémente ? On devrait retrouver dans la description de la PR, le fait de rajouter "sources" à query, que c'est quelque chose qui est pas dans le DSL mais adhoc pour le moment, etc...

Reply:
Ok je vois. L'idée de la 404 c'était aussi de se dire que la cause est côté utilisateur (problème internet) et donc que dans ce cas, inutile de faire une deuxième requête.

Reply:
En cas de "problème internet", on recevra même pas de 404, on aura une exception.